# Exploring eBay Car Sales Data

In this project, we will be working with a dataset of used cars from eBay Kleinanzeigen, a [classified](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. Originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded by user [orgesleka](https://www.kaggle.com/orgesleka), the dataset is no longer available on Kaggle but can be found [here](https://data.world/data-society/used-cars-data). The dataset has also been dirtied and modified to hold 50,000 data points from the full dataset.

As we explore the "autos.csv" dataset, we will clean up any incorrect data or errors. We will also make some initial observations of what we find. Our goal is to illustrate the usefulness of data analysis when dealing with substantial amounts of data from any field or business.

To start, let's import our libraries, and start exploring our dataset.

#### Opening Libraries & Initial Exploraton

In [2]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='latin1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In exploring our dataset above, using the `.info()` and `.head()` methods, we have found that our dataset contains 50000 rows and 20  columns. We observe the data type and "Non-Null Count" for each column. Columns with a "Non-Null Count" lower than 50000 indicate the presence of null values in that column. We will address these columns later, as null values can sometimes pose problems for analysis.

Next, we have noticed that the column names are in [camelcase](https://en.wikipedia.org/wiki/Camel_case) format. In the code below, we will convert all the column names from the camelcase format to the [snakecase](https://en.wikipedia.org/wiki/Snake_case) format, which is more commonly used in Python.

#### Column Index Correction

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
mapping_dict = {
    'dateCrawled': 'date_crawled',
    'name': 'name',
    'seller': 'seller',
    'offerType': 'offer_type',
    'price': 'price',
    'abtest': 'ab_test',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'gearbox': 'gearbox',
    'powerPS': 'powerPS',
    'model': 'model',
    'odometer': 'odometer',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'brand': 'brand',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'nr_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen',
}
autos.columns = autos.columns.map(mapping_dict)

In [8]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Above we changed most of the column names simply by adding an underscore in place of the space and making the first letter of the second word lowercase. However, there were a few column names that we changed more drastically to be more descriptive or concise. These  extreme changes were:

- yearOfRegistration : registration_year
- monthOfRegistration : registration_month
- notRepairedDamage : unrepaired_damage
- dateCreated : ad_created

Now that we have corrected the column names, we will now move on to a deeper exploration of the data.

#### Exploration

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In the results above, there are several columns with curious data:

- The columns with the majority of the values being the same:
    - `seller` - only 2 unique values
    - `offer_type` - only 2 unique values
- Odd columns:
    - `nr_of_pictures` - the results above for this column were either 'NaN' or 0.0
- Numeric values stored as text:
    - `price`
    - `odometer`
- Absurd max and min:
    - `registration_year` - a car cannot be registered after it is listed

Now that we have identified some errors, we can begin cleaning our dataset. We'll start by looking more closely at the `seller`, `offer_type`, and `nr_of_pictures` columns. In the `seller` and `offer_type` columns, we are looking to see if they hold any useful information since they each have only 2 unique values. As for the `nr_of_pictures` column, we are unsure of its value as well, since its results above were unusual.

#### `seller`, `offer_type`, `nr_of_pictures` - Close Examination

In [10]:
print(autos.iloc[:, 2].value_counts())
print("\n")
print(autos.iloc[:, 3].value_counts())
print("\n")
print(autos.iloc[:, 17].value_counts())

seller
privat        49999
gewerblich        1
Name: count, dtype: int64


offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64


nr_of_pictures
0    50000
Name: count, dtype: int64


Looking at the results above, we can see that all three of these columns do not hold any useful information. With this conclusion, we shall proceed to drop the unneeded columns.

#### Dropping the `seller`, `offer_type`, and `nr_of_pictures` columns

In [11]:
autos = autos.drop(['seller'], axis=1)

autos = autos.drop(['offer_type'], axis=1)

autos = autos.drop(['nr_of_pictures'], axis=1)

With the `seller`, `offer_type`, and `nr_of_pictures` columns taken care of, we can now move on to the `price` and `odometer` columns. Let's study these two columns a bit more closely before deciding our next move.

#### `price` and `odometer` - Close Examination

In [12]:
print(autos.iloc[:, 2].value_counts())
print("\n")
print(autos.iloc[:, 9].value_counts())

price
$0            1421
$500           781
$1,500         734
$2,500         643
$1,000         639
              ... 
$35,222          1
$34,490          1
$910             1
$3,890,000       1
$48,600          1
Name: count, Length: 2357, dtype: int64


odometer
150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: count, dtype: int64


As we can see, the data in these two columns contain the "$" "," and "km" non-numeric characters. These characters can confuse our calculations when we form our analysis. However, these characters cannot be removed from the data entirely, as they contain important information about the measurement or structure of the data within the column. To avoid this, we can also rename the columns to include the measurement values.

#### Converting `price` and `odometer` to numeric data types

In the code below, we will iterate through the `price` and `odometer` columns and remove any non-numeric characters. After the columns are clean we will convert them to a numeric data type. Finally, we will rename each column to include the measurement type.

In [13]:
autos["price"] = (autos["price"]
                  .str.replace('$','')
                  .str.replace(',','')
                  .astype(int)
                 )
autos.rename({"price": "price_$"}, axis=1, inplace=True)
autos["price_$"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_$, dtype: int64

In [14]:
autos["odometer"] = (autos["odometer"]
                  .str.replace(',','')
                  .str.replace('km','')
                  .astype(int)
                 )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Before we move on, let's look a bit deeper into our new and improved `price_$` and `odometer_km` columns. This time we will be looking for any outlying data, data that is unrealistically high or low. In the code below, we will be using several different methods such as `.unique().shape` and `.describe()` to closely inspect our data.

#### `price_$` - Close Examination

In [15]:
print("How many unique values:")
print(autos["price_$"].unique().shape)
print("\n")
print("To view min/max/median/mean etc:")
print(autos["price_$"].describe())
print("\n")
print("To view the highest and lowest values with their counts:")
print(autos["price_$"].sort_index(ascending=False))

How many unique values:
(2357,)


To view min/max/median/mean etc:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


To view the highest and lowest values with their counts:
49999     1250
49998    22900
49997    13200
49996     1980
49995    24900
         ...  
4         1350
3         4350
2         8990
1         8500
0         5000
Name: price_$, Length: 50000, dtype: int64


In studying our results above, we can see that we have 2357 unique values in the `price_$` column. This could be due to prices being rounded on the website. In looking at our lower range values we notice a large number of '0' values. Since listings on the German eBay website are sold in the bidding style, it is possible for a bid to start at \\$1. However, since there cannot be a bid or sale of \\$0, we will end up dropping the rows containing a `0` price value.

Before we start dropping rows, let's take a closer look at the higher range of values as well. In the results above we notice a maximum value of \$100 million, which seems unrealistically high, and deserves a deeper look.

In [16]:
autos["price_$"].value_counts().sort_index(ascending=False).head(20)

price_$
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64

As we can see, there are some extremely high values here. They seem to be inconsistent until they get to "350000", where they gradually descend from there. Since all the values above "350000" are so extreme, we will drop those rows as well.

#### Removing Outlying Rows Based On `price_$` column

From our observations of both the low and high ranges of prices, we have decided to keep only the rows within the price range [1 - 350000]. In the code below we will remove all the rows outside that range.

In [17]:
autos = autos[autos["price_$"].between(1,350000)]
autos["price_$"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

Now that we have taken care of the price column, let's turn our attention to the `odometer_km` column. Just like with the price column, we are looking for data that is out of place. We will be performing similar exploration methods as we did with the `price_$` column.

#### `odometer_km` - Close Examination

In [18]:
print("How many unique values:")
print(autos["odometer_km"].unique().shape)
print("\n")
print("To view min/max/median/mean etc:")
print(autos["odometer_km"].describe())
print("\n")
print("To view the values with their counts:")
autos["odometer_km"].value_counts()

How many unique values:
(13,)


To view min/max/median/mean etc:
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


To view the values with their counts:


odometer_km
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: count, dtype: int64

In our exploration results, we notice a higher concentration of values in the higher range compared to the lower range. Also, all the values are rounded numbers, indicating that there may have been a list of odometer value options to select from when listing a vehicle. Other than that, there seem to be no outliers or out-of-place data.

#### Exploring the Date Columns

Let's move on to the date columns. The five columns that should represent date values are:

- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

In the code below, we will begin by looking at the data types for each of these columns.

In [19]:
print("date_crawled: " + str(autos["date_crawled"].dtype) + "\n" + 
      "last_seen: " + str(autos["last_seen"].dtype) + "\n" + 
      "ad_created: " + str(autos["ad_created"].dtype) + "\n" + 
      "registration_month: " + str(autos["registration_month"].dtype) + "\n" + 
      "registration_year: " + str(autos["registration_year"].dtype))

date_crawled: object
last_seen: object
ad_created: object
registration_month: int64
registration_year: int64


We can see that the columns not holding registration information are all identified as string data types. To continue our exploration, we will look into the format of these string columns in the code below.

In [20]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Above, we notice that the first ten characters of each entry represent the date. Let's use this information to help us understand the date range. Next, we will use the first ten characters in each column to show us the date range.

In [21]:
print(autos['date_crawled'].str[:10])
print("\n")
print(autos['ad_created'].str[:10])
print("\n")
print(autos['last_seen'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 48565, dtype: object


0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 48565, dtype: object


In the result above, we can see a preview of the date values (first ten characters) in each column. We will now use the `Series.value_counts(normalize=True, dropna=False).sort_index()` method to describe the distribution of values as percentages for each column.

In [22]:
# distribution of values in the date_crawled column as percentages
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

In the `date_crawled` column, we can see that the date with the highest percentage is April 3, 2016. We also notice that while the percentages of the day vary, all the values are either in March or April of 2016. We will make a note of this, and move on to the next column, `ad_created`.

In [23]:
# distribution of values in the ad_created column as percentages
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64

Again we notice that the highest percentages of dates in the `ad_created` column are in March (03) and April (04). However, unlike the `date_crawled` column, several entries are much older. The older entries seem to date back as far as nine months.

Let's move on to our final column in this exploration step, the `last_seen` column.

In [24]:
# distribution of values in the last_seen column as percentages
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64

With these results, we can see that the March and April pattern continues. Since we notice that these dates also seem to be the highest, we can begin to presume that these may have been the few final dates the site was crawled. This is more likely than an unexpected increase in transactions.

We now have seen the date range in the `date_crawled`, `ad_created`, and `last_seen` columns, which go from about June 2015 to April 2016. Let's move on to understand the distribution of the `registration_year` column. This column would be most useful to us for analysis purposes because it has a date range, similar to the `date_crawled`, `ad_created`, and `last_seen` columns. Since the data in the `registration_month` column is in months and not years, we will not end up exploring it further.

In [25]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

One thing we notice right away from this result is that our minimum and maximum values seem to be very strange. If you recall, the latest year in the listing values was 2016. Since a car can't be registered after it is listed, any year over 2016 is incorrect. Furthermore, there can't be a vehicle registered in the year 1000. It seems we have some cleaning to do.

To clean the dataset of incorrect rows, we need to create a range containing the accurate years. From our above observations, we have already identified 2016 as the maximum year in our range. If we brush up on the [history](https://www.history.com/topics/inventions/automobiles) of automobiles, we see that although they were first invented in the 1800s, they weren't commonly used among the public until the 1900s. Based on this information, we will write our range from `1900` to `2016`.

Before we remove the rows outside our range, we need to make sure it is safe to do so. Let's do a bit more investigating in the code below.

In [26]:
autos['registration_year'].value_counts(normalize=True).sort_index()

registration_year
1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: proportion, Length: 95, dtype: float64

The above results show that the percentage of values outside our range is very small. Therefore we now know it is safe to drop them. In the code below, we will keep only the rows within our range.

In [27]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

We have now removed the incorrect rows based on our range from the `registration_year` column. 

#### Exploring Price by Brand

Let's move on to explore the `brand` column.

In [28]:
autos['brand'].describe()

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [29]:
autos['brand'].value_counts(normalize=True)

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

Above, we can observe the percentage of appearances for each brand in the column. We also notice that the top few brands have significantly higher percentages than the others. Other than these initial observations, there doesn't appear to be anything wrong with the data in this column.

We will now calculate the average price for a range of specific brands. To start, we need to select the brands. Since we are not going to calculate the average price for all brands, it would be best to select only those that are large enough to support an accurate average. For this challenge, we will select the brands that account for more than 5% of all the listings.

In the code below we will create a dictionary using a `for` loop. The dictionary will contain each brand from our selections, and its corresponding average price.

In [30]:
# separating out our selected brands

brand_percents = autos["brand"].value_counts(normalize=True)
selected_brands = brand_percents[brand_percents > .05].index
print(selected_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In [31]:
# finding the average price for each brand
avg_price_by_brand = {}

for b in selected_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["price_$"].mean()
    avg_price_by_brand[b] = int(mean)

print("\n".join("{0}: {1}".format(k, v)  for k,v in avg_price_by_brand.items()))

volkswagen: 5402
bmw: 8332
opel: 2975
mercedes_benz: 8628
audi: 9336
ford: 3749


In [32]:
# sorting the 'avg_price_by_brand' dictionary from highest price to lowest
avg_brand_price_sorted = dict(sorted(avg_price_by_brand.items(), key=lambda x:x[1], reverse=True))
print("\n".join("{0}: {1}".format(k, v)  for k,v in avg_brand_price_sorted.items()))

audi: 9336
mercedes_benz: 8628
bmw: 8332
volkswagen: 5402
ford: 3749
opel: 2975


From our results above, we can see that there are evident gaps between the prices:

- Audi, BMW, and Mercedes Benz range high
- Ford and Opel range low
- Volkswagen is in the middle

#### Exploring Mileage by Brand

Now let's do the same thing we did before, but in this case, calculate the average mileage for our six selected brands.

In [33]:
# finding the average mileage for each brand
avg_brand_mileage = {}

for b in selected_brands:
    selected_rows = autos[autos["brand"] == b]
    mean = selected_rows["odometer_km"].mean()
    avg_brand_mileage[b] = int(mean)

print("\n".join("{0}: {1}".format(k, v)  for k,v in avg_brand_mileage.items()))

volkswagen: 128707
bmw: 132572
opel: 129310
mercedes_benz: 130788
audi: 129157
ford: 124266


In [34]:
# sorting the 'avg_brand_mileage' dictionary from highest mileage to lowest
avg_brand_mileage_sorted = dict(sorted(avg_brand_mileage.items(), key=lambda x:x[1], reverse=True))
print("\n".join("{0}: {1}".format(k, v)  for k,v in avg_brand_mileage_sorted.items()))

bmw: 132572
mercedes_benz: 130788
opel: 129310
audi: 129157
volkswagen: 128707
ford: 124266


Here, we can see that the BMW brand has the most mileage, while the others have incrementally less.

#### Average Price and Average Mileage Comparison

We now have the average price and average mileage information about our six brands. This causes us to raise the question of whether they could be related. Will a brand that is priced lower, also have higher mileage? Or are the more expensive brands more popular?

In the code below, we will put our two dictionaries in a data frame format, to compare them easier.

In [35]:
price_series = pd.Series(avg_brand_price_sorted)
mileage_series = pd.Series(avg_brand_mileage_sorted)

comparison_df = pd.DataFrame(price_series, columns=['average_price'])
comparison_df['average_mileage'] = mileage_series

print(comparison_df)

               average_price  average_mileage
audi                    9336           129157
mercedes_benz           8628           130788
bmw                     8332           132572
volkswagen              5402           128707
ford                    3749           124266
opel                    2975           129310


As we observe our new data frame, we notice some interesting relationships:
- BMW is priced close to the middle, while its mileage is the highest
- Mercedes Benz has the second highest price and the second highest mileage
- Opel is the lowest price with the third highest mileage
- Audi is priced the highest, while its mileage is in the lower range, fourth highest
- Volkswagen is priced in the middle range and has the second-lowest mileage
- Ford has the second-lowest price and has the lowest mileage

While the average mileage doesn't vary widely, it's intriguing to observe the relationships between each brand's price and mileage. Although we might have expected lower-priced brands to have higher mileage, that is not what these results showed. Instead, we can see a small correlation between price and mileage, with higher-priced brands having higher mileage.

#### Conclusion

Throughout this project, we have made some valuable observations as we explored. We have had opportunities to identify errors and correct them. We have analyzed data trends and discovered that the results are not always as expected. While there is so much more that could be accomplished with this dataset, this project demonstrates the value and productivity that data analysis can provide to any industry.